### PR TITLE
Bug 1502921 - Add new 'intl' fields to environment data schema

### PR DESCRIPTION
**Pull request by flodolo**
CC @mstriemer

Bug 1502921 added a few field to the Environment ping. Per IRC discussion, we need this data to be validated in order to add it to datasets in STMO.

Landing in m-c (Firefox 66): https://hg.mozilla.org/mozilla-central/rev/de6426e0f491

### CODE REVIEW

File: templates/include/telemetry/environment.1.schema.json [hunk 1]
```diff
@@ -357,6 +357,29 @@
         "e10sCohort": {
           "type": "string"
         },
+        "intl": {
+          "type": "object",
+          "properties": {
+            "requestedLocales": {
+              "type": "string"
```
Comment (mstriemer) [inline templates/include/telemetry/environment.1.schema.json]: I think these should be:

```json
"requestedLocales": {
  "type": "array",
  "items": {
    "type": "string"
  }
```

Comment (flodolo) [reply templates/include/telemetry/environment.1.schema.json]: Ah, I missed we have an array type. I think you're right

Comment (flodolo) [reply templates/include/telemetry/environment.1.schema.json]: I force pushed since GH these days shows them anyway in the history

**Comment by mstriemer** [comment]
`regionalPrefsLocales` and `systemLocales` can be null instead of an array. This is similar to `locale` which can be a string or null, but that only defines `"type": "string"`. Not sure if anything needs to be done to handle that.

They shouldn't be null, but the code allows for it if the mozIOSPreferences object thinger isn't loaded for some reason (likely just on android tests, but I don't know for sure).

**Comment by flodolo** [comment]
> `regionalPrefsLocales` and `systemLocales` can be null instead of an array. This is similar to `locale` which can be a string or null, but that only defines `"type": "string"`. Not sure if anything needs to be done to handle that.

Would that be this?

```
"regionalPrefsLocales": {
    "type": [ "array", "null" ],
    "items": {
    "type": "string"
    }
},
```

**Comment by mreid-moz** [comment]
> I think so, but I'd defer to a maintainer of this project. Seeing as it isn't set for locale I'm not sure it's needed.

It would be a good idea to add a couple of examples to the validation pings that exercise this case. If the field is simply missing in the JSON, you don't need to include `null` as a type, but if you see a literal `"regionalPrefsLocales": null` in the incoming documents, it _is_ needed.

**Comment by flodolo** [comment]
> It would be a good idea to add a couple of examples to the validation pings that exercise this case

Mark, would you be able to provide that?

**Comment by acmiyaguchi** [comment]
@flodolo To clarify Mark's comment, it's good practice to add validation pings that functionally act as unit tests. They can be added to `mozilla-pipeline-schemas/validation/telemetry`. 

The convention is `<document-type>.<version>.<validation-name>.<[pass | fail]>.json`. You can start with the minimal main document called [`main.4.min.pass.json`](https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/dev/validation/telemetry/main.4.min.pass.json) to construct the test cases:

https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/8939bcb719855c6d84fbf2cfb4c256f2123d7026/validation/telemetry/main.4.min.pass.json#L1-L16

**Comment by flodolo** [comment]
> @flodolo To clarify Mark's comment, it's good practice to add validation pings that functionally act as unit tests. They can be added to `mozilla-pipeline-schemas/validation/telemetry`.

Would this be run automatically via integration, or do I need to run tests manually (and how)?

I'd still need some input from Mark on the null cases, because I'm not really a developer, and I'm not sure which cases should be tested.

**Comment by acmiyaguchi** [comment]
They are run during integration, you can "Show all checks" to see the logs from the tests. If you have docker (or have an environment with cmake installed), you can follow the [instructions in the README](https://github.com/mozilla-services/mozilla-pipeline-schemas#running-tests-via-docker) to test locally.

It looks like there are three cases that should be tested from this thread, to be comprehensive:

1. The normal case where the field is an array (new pings)
2. A degenerate case where the field is null per https://github.com/mozilla-services/mozilla-pipeline-schemas/pull/248#issuecomment-457334336
3. A case where the field doesn't exists (for pre-existing pings)

Using this schema as reference:

```
"regionalPrefsLocales": {
    "type": [ "array", "null" ],
    "items": {
        "type": "string"
    }
},
```

You might have something like:

```json
"intl": {
    ...
}
```

```json
"intl": {
    ...
    "regionalPrefsLocales": null,
    ...
}
```

and

```json
"intl": {
    ...
    "regionalPrefsLocales": ["hello", "world"],
    ...
}
```

**Comment by flodolo** [comment]
Thanks for all the info. I'll take a look next week and see what I can come up with.

**Comment by flodolo** [comment]
I think I finally figured out how to get tests pass…

@mstriemer 
Can you verify if the test cases make sense based on the code?

**Comment by flodolo** [comment]
@mstriemer review ping :)

Side question: if/when this is merged, what are the next steps? Is the data going to be backfilled, since pings already include this information?

**Comment by mstriemer** [comment]
The schema and tests look right to me.

**Comment by flodolo** [comment]
Sorry, somehow your reply got completely lost in my mailbox.

I rebase on top of dev just in case, run the command and added it as a separate commit.

**Comment by acmiyaguchi** [comment]
@flodolo Thanks for updating the schema :+1:. Adding this information to the schema will allow us to validate the pings in the pipeline. There's no backfill necessary, since extra fields are generally allowed in the documents. 

The data can be access through the pings via the [moztelemetry API](https://docs.telemetry.mozilla.org/tools/spark.html#the-moztelemetry-library) in a python notebook. If you want to query using sql.telemetry.mozilla.org, the new environment fields  would need to be added explicitly to a dataset like `main_summary`.

**Comment by flodolo** [comment]
> If you want to query using sql.telemetry.mozilla.org, the new environment fields would need to be added explicitly to a dataset like `main_summary`.

How can that be done? All the queries and dashboards I have live on STMO, so it would be great to have that. 

When adding these fields we were also considering removing the `locale` field, but left it alone as a future follow-up.

**Comment by flodolo** [comment]
(For further reference: I normally look at the `client_count_daily` dataset)

**Comment by acmiyaguchi** [comment]
The best next step is to file a bug under `Data Platform and Tools :: Datasets: Main Summary` to add the new environment fields to the dataset. We wouldn't do a backfill on the data here, since the fields haven't hit the release channel yet, and it's time consuming to backfill.

**Pull request merged**